### PR TITLE
Implement FindCommand to search using prefixes

### PR DESCRIPTION
**Pull request by Gavino3o**
# Changes
Previously, the `find` command can only be used to find patients or specialists by their `Name` using full String comparison. 
The changes in this PR will modify the `find` command so that it will now accept prefixes (similar to `add` command, `edit` command etc) and return any person satisfying **_ALL_** find conditions. The prefixes are optional.

### Notes: 
- With this new implementation, if a `find` command is executed without any prefixes, it behaves like a `list` command.
- For multivalued attributes of a person (e.g. `Tags`), the `find` command will return any person with at least one of the values of the attribute. For example: Patient Bob has the tags: 'friends' and 'colleague', `find -pa t/friends` or `find -pa t/colleague` will return Bob as one of the results.
- As of now, each attribute is being searched using Full String Comparison of Case-Insensitive White Spaced Words in the original arguments. This can be improved to use substring search in the future.
- Test for added changes will be added in a future increment.

### CODE REVIEW

**Comment by Gavino3o** [comment]
Test for added changes will be added in a future increment.

File: src/main/java/seedu/address/logic/parser/FindCommandParser.java [hunk 1]
```diff
@@ -20,15 +37,92 @@ public class FindCommandParser implements ParserComplex<FindCommand> {
...
+
+    private FindCommand parsePatient(String args) throws ParseException {
+        ArgumentMultimap argMultimap =
+                ArgumentTokenizer.tokenize(args, PREFIX_NAME, PREFIX_PHONE, PREFIX_EMAIL, PREFIX_ADDRESS,
+                        PREFIX_TAG, PREFIX_MEDICALHISTORY);
+
+        argMultimap.verifyNoDuplicatePrefixesFor(PREFIX_NAME, PREFIX_PHONE, PREFIX_EMAIL,
+                PREFIX_ADDRESS, PREFIX_MEDICALHISTORY);
+
+        List<Predicate<Person>> predicateList = setupPersonPredicates(argMultimap);
```
Comment (alvinlim277) [inline src/main/java/seedu/address/logic/parser/FindCommandParser.java:60]: Really like the abstraction of `setupPersonPredicates` here, makes the function a lot easier to understand.

File: src/main/java/seedu/address/model/person/predicates/TagsContainsKeywordsPredicate.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+    public TagsContainsKeywordsPredicate(List<String> keywords) {
+        this.keywords = keywords;
+    }
+
+    @Override
+    public boolean test(Person person) {
+        return person.getTags().stream().map(tag -> keywords.stream()
+                    .anyMatch(keyword -> StringUtil.containsWordIgnoreCase(tag.tagName, keyword)))
+                    .reduce(false, (x, y) -> x || y);
```
Comment (alvinlim277) [inline src/main/java/seedu/address/model/person/predicates/TagsContainsKeywordsPredicate.java:25]: Very nice use of streams and lambda functions. Perhaps the readability of this complex stream manipulation could be improved with better used of newlines and indentation.
For example, the outer stream(), map() and reduce() should be on the same indentation level, while the inner stream() and anyMatch() can be on a separate indentation level.

**Review by alvinlim277 — COMMENTED** [review]
LGTM, just a minor nit regarding readability. Overall great implementation of the FindCommand that makes it a lot more versatile and user-friendly! The robust OOP and stream manipulation is very impressive.

File: src/main/java/seedu/address/model/person/predicates/TagsContainsKeywordsPredicate.java (hunk 2 above)
Comment (Gavino3o) [reply src/main/java/seedu/address/model/person/predicates/TagsContainsKeywordsPredicate.java]: Thank you for pointing that out, I completely agree! I have made the changes.

**Review by alvinlim277 — APPROVED** [review]
LGTM!

**Pull request merged**